tachyon/zk/plonk/circuit/selector_compressor.h

```diff
@@ -165,61 +221,67 @@ class SelectorCompressor {
-      size_t combination_index = combination_assignments.size();
-      std::unique_ptr<Expression<F>> query = callback.Run();
-
-      F assigned_root = F::One();
-      selector_assignments.reserve(selector_assignments.size() +
-                                   combination.size());
-      for (const SelectorDescription& selector : combination) {
-        // Compute the expression for substitution. This produces an expression
-        // of the form
-        //     q * Prod[i = 1..=combination_len, i != assigned_root](i - q)
-        //
-        // which is non-zero only on rows where |combination_assignment| is set
-        // to |assigned_root|. In particular, rows set to 0 correspond to all
-        // selectors being disabled.
-        std::unique_ptr<Expression<F>> expression = query->Clone();
-        F root = F::One();
-        for (size_t i = 0; i < combination_len; ++i) {
-          if (root != assigned_root) {
-            expression = std::move(expression) *
-                         (ExpressionFactory<F>::Constant(root) - query);
-          }
-          root += F::One();
-        }
+      ConstructCombinedSelector(n, combination);
+    }
+  }
 
-        // Update the combination assignment
-        const std::vector<bool>& activations = selector.activations();
-        for (size_t k = 0; k < n; ++k) {
-          // This will not overwrite another selector's activations
-          // because we have ensured that selectors are disjoint.
-          if (activations[k]) {
-            combination_assignment[k] = assigned_root;
-          }
+  void ConstructCombinedSelector(
+      size_t n, const std::vector<SelectorDescription>& combination) {
+    // Now, compute the selector and combination assignments.
+    std::vector<F> combination_assignment = base::CreateVector(n, F::Zero());
+    size_t combination_len = combination.size();
+    size_t combination_index = combination_assignments_.size();
+    std::unique_ptr<Expression<F>> query = callback_.Run();
+
+    F assigned_root = F::One();
+    selector_assignments_.reserve(selector_assignments_.size() +
+                                  combination.size());
+    for (const SelectorDescription& selector : combination) {
+      // Compute the expression for substitution. This produces an expression
+      // of the form
+      //     q * Prod[i = 1..combination_len, i != assigned_root](i - q)
+      //
+      // which is non-zero only on rows where |combination_assignment| is set
+      // to |assigned_root|. In particular, rows set to 0 correspond to all
+      // selectors being disabled.
+      // For example, in the first iteration, it produces:
+      //     q * (2 - q) * (3 - q)
+      std::unique_ptr<Expression<F>> expression = query->Clone();
+      F root = F::One();
+      for (size_t i = 0; i < combination_len; ++i) {
+        if (root != assigned_root) {
+          expression = std::move(expression) *
+                       (ExpressionFactory<F>::Constant(root) - query);
```

Comment:
It seems that expression is always computed with the `1 - q`! And we can optimize by removing duplicated parts.